system/CodeIgniter.php

```diff
@@ -345,13 +356,22 @@ public function run(?RouteCollectionInterface $routes = null, bool $returnRespon
             // If the route is a 'redirect' route, it throws
             // the exception with the $to as the message
             $this->response->redirect(base_url($e->getMessage()), 'auto', $e->getCode());
-            $this->sendResponse();
+
+            if (! $this->returnResponse) {
+                $this->sendResponse();
+            } else {
+                return $this->response;
+            }
```

Comment:
this can be changed to:

```suggestion
            if ($this->returnRespose) {
                return $this->response;
            }

            $this->sendResponse();
```